File: linked-list/reverseList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null) return null;
        if (head.next == null) return head;
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode back = dummy;
        ListNode current = head;
        ListNode front = head.next;
        while (current.next != null){
            current.next = front.next;
            if (back == dummy) {
                front.next = head;
                back = front;
            }
            else{
                front.next = back;
                back = front;
            }
            
            front = current.next;
        }
        return back;
    }
}

File: recursion/reverseList.py
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        pre = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return pre
      
     def reverseList(self, head: ListNode) -> ListNode:        
        pre = None
        curr = head
        while curr:
            tmp = curr.next
            curr.next = pre
            pre = curr
            curr = tmp
        return pre
